mcs/class/corlib/System.IO/LogcatTextWriter.cs

```diff
@@ -48,22 +48,52 @@ public override void WriteLine ()
 			var o = line.ToString ();
 			line.Clear ();
 
+			Log (o);
+			stdout.WriteLine (o);
+		}
+
+		void Log (string message)
+		{
+			const int buffer_size = 512;
+
 			unsafe {
-				fixed (byte *b_appname = appname)
-				fixed (byte *b_message = Encoding.UTF8.GetBytes(o)) {
-					Log (b_appname, 1 << 5 /* G_LOG_LEVEL_MESSAGE */, b_message);
+				if (message.Length < buffer_size) {
+					try {
+						fixed (char *b_message = message) {
+							byte* buffer = stackalloc byte[buffer_size];
+							int written = Encoding.UTF8.GetBytes (b_message, message.Length, buffer, buffer_size - 1);
+							buffer [written] = (byte)'\0';
+
+							Log (buffer);
+						}
+
+						return;
+					} catch (ArgumentException) {
+						/* It might be due to a failure to encode a character, or due to a smaller than necessary buffer. In the
+						* secode case, we want to fallback to simply allocating on the heap. */
+					}
+				}
+
+				fixed (byte *b_message = Encoding.UTF8.GetBytes(message + '\0')) {
+					Log (b_message);
 				}
 			}
-			stdout.WriteLine (o);
 		}
 
-		public static bool IsRunningOnAndroid ()
```

Comment:
In terms of the diff itself, it's odd that you moved `IsRunningOnAndroid()` to be *after* all the `Log()` overloads. The patch would be smaller/have less noise if the `Log(byte*)` overload were *after* the "existing" `IsRunningOnAndroid()`.